[DOC] Cruise Control: rebalance procedures and overview of goals and proposals

laidan6000 opened the pull request:
### Type of change

- Documentation

### Description

Added three new modules to the existing _Cruise Control_ section of the _Using Strimzi_ guide.

- "Rebalance goals and proposals" (concept)
- "Generating rebalance proposals" (procedure)
- "Implementing a rebalance proposal" (procedure)

To make the content easier to review, I recommend using the Asciidoctor.js Live Preview extension for [Chrome](https://chrome.google.com/webstore/detail/asciidoctorjs-live-previe/iaalpfgpbocpdfblpnhhgllgbdbchmia) or [Firefox](https://addons.mozilla.org/fr/firefox/addon/asciidoctorjs-live-preview/). You can simply open the raw file in a new browser tab and the extension will render the markup for you.

Also reordered the `include` statements in the assembly. Hopefully the information is introduced in a logical order!

Made some minor edits to the existing Cruise Control documentation. However, future PRs will need to make further edits to align the terminology, e.g. "rebalance proposals" instead of "optimization proposals".

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [X] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [X] Reference relevant issue(s) and close them after merging - NO ISSUES FOUND
- [ ] Update CHANGELOG.md

ppatierno commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+They are organized in a customizable hierarchy, from high priority goals to low priority goals. 
+You can designate certain rebalance goals as _hard goals_ that must be satisfied--for more information, see link:#soft-hard-goals[Soft goals and hard goals].
+
+With a few exceptions, cluster rebalancing in {ProductName} supports the same goals as the Kafka Cruise Control project (where the term "optimization goals" is used). 
```
What does it mean "...supports the same goals as the Kafka Cruise Control"? Strimzi is using Kafka Cruise Control! Thinking on this, if the CC community names them "optimization goals" why we should change to "rebalance goals"? Isn't it better having consistency?

laidan6000 replied:
I was under the impression that Strimzi will not support _all_ of the CC optimization goals. Is that not the case?

Re. "rebalance goals" vs "optimization goals", the integration with Cruise Control is for cluster rebalancing, not the other CC features such as Anomaly Detection, Topic Replication Factor. Therefore, we felt that "rebalance goals" was a better name.

But I'm happy to rename to "optimization goals" -  @tomncooper, do you have any strong opinions?

tomncooper replied:
Yes, optimization goals is a better name, not least because eventually we may well support most of the CC features.

laidan6000 replied:
Good point. Time to Find and Replace... ;)

ppatierno commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+
+NOTE: "Write your own" goals, Kafka assigner goals, and Anomaly Detector are not supported in {ProductName}.
+
+For detailed information on each rebalance goal, see link:https://github.com/linkedin/cruise-control/wiki/Pluggable-Components#goals[Goals^] in the Cruise Control Wiki. 
```
I would swap the order between this line and the NOTE above.

ppatierno commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+
+¦Also known as "best effort" goals, soft goals do _not_ need to be satisfied in rebalance proposals. 
+However, they are included in rebalance calculations. 
+A rebalance proposal that satisfies _all_ hard goals but violates one or more soft goals is valid.
```
because we are talking about soft goals, can we swap the order in this sentence, something like ..." A rebalance proposal that violates one or more soft goals but satisfies all hard goals is valid"

laidan6000 replied:
Nice suggestion, thanks.

ppatierno commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+¦Hard goals
+
+¦Hard goals _must_ be satisfied in rebalance proposals. 
+In other words, a rebalance proposal that does _not_ satisfy all the designated hard goals is rejected by Cruise Control and not sent to the user.
```
even if in a rebalance proposal request you can set a flag like "skip hard goals check" and in this case, should we mention it here? wdyt @tomncooper @kyguy ?

tomncooper replied:
Yes and link to the custom goals documentation?

kyguy replied:
Even though we don't support custom goals?

kyguy replied:
Ah you mean the goal shipped with CC, not write your own goals [1]

[1] https://github.com/linkedin/cruise-control/wiki/Write-your-own-goals

laidan6000 replied:
Rephrased as follows:

"Hard goals _must_ be satisfied in rebalance proposals unless the `skip_hard_goal_check` option is specified."

@tomncooper - I can't find any suitable information in the Cruise Control wiki.

ppatierno commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+[discrete]
+=== Default and custom rebalance goals
+
+When interacting with Cruise Control to generate rebalance proposals, you can use the _default rebalance goals_, which are defined on deployment, or set ad hoc _custom rebalance goals_. 
```
"are defined in the cruise control deployment configuration"

tomncooper replied:
Can we call them optimization proposals rather than rebalance proposals.

ppatierno commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+
+*Custom rebalance goals* 
+
+Custom rebalance goals are set in the `spec.goals` section of a `KafkaClusterRebalance` resource and override the default rebalance goals. 
```
now the resource is `KafkaRebalance`. Didn't we decide to not use the word "custom" here because it conflicts with the custom goal classes that a developer can write by himself?

laidan6000 replied:
Will update to `KafkaRebalance` throughout.

I thought we decided using "custom goals" was OK! We need to use some kind of qualifier to help readers understand how they differ from default goals. Suggestions:

- Per operation goals
- Per rebalance goals
- Ad hoc goals
- Operation-specific goals
- Rebalance-specific goals
- User goals

ppatierno replied:
I am for "user-provided goals" (where provided means that they are provided through a `KafkaRebalance` resource)

PaulRMellor replied:
I might be getting a bit confused. If there are default, user-provided and custom goals, we need to make sure that the reader understands what each mean up front.

laidan6000 replied:
Happy to use "user-provided goals", which I think makes sense in context.

@PaulRMellor - I'll have another think about how to introduce the information. There are two types of goal -- default and user-provided -- that can be configured to include hard and soft goals.

ppatierno commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+
+Use custom rebalance goals to generate a rebalance proposal that addresses a particular optimization scenario. 
+For example, you might want to optimize leader replica distribution across the Kafka cluster without considering other goals, such as disk or CPU usage. 
+So, you create a `KafkaClusterRebalance` resource containing a custom rebalance goal for leader replica distribution only.
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+A _rebalance proposal_ is a summary of proposed changes from Cruise Control that would produce a more balanced Kafka cluster, with partition workloads distributed more evenly across the brokers. 
+Each rebalance proposal is based on the set of rebalance goals (either default goals or custom goals) that were used to generate it.
+
+Rebalance proposals always satisfy all the supplied hard goals, though not necessarily all the supplied soft goals.
```
Again you can set a skip hard goals check flag in the `KafkaRebalace` custom resource.

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,98 @@
+[id='proc-generating-rebalance-proposals-{context}']
+= Generating rebalance proposals 
+
+When you create or update a `KafkaClusterRebalance` resource, Cruise Control generates a _rebalance proposal_ in JSON.
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,98 @@
+
+.Procedure
+
+. Create a `KafkaClusterRebalance` resource:
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,98 @@
+[source,yaml,subs="attributes+"]
+----
+apiVersion: {KafkaApiVersion}
+kind: KafkaClusterRebalance
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,98 @@
+[source,yaml,subs="attributes+"]
+----
+apiVersion: {KafkaApiVersion}
+kind: KafkaClusterRebalance
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,98 @@
+spec:    
+    goals: >
+      com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,
+      com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal
```
The FQDN is not needed, just the classes names.

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,98 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl get kafka kafkaclusterrebalance -o jsonpath='{.status}'
```
`kafkarebalance`

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,98 @@
+----
+Cruise Control returns one of two statuses:
+
+** `PendingProposal`: The rebalance proposal is not yet ready; try again after a few minutes.
```
no, there is no action to take by the user in a few minutes. The rebalance operator is polling the cruise control API to check when the proposal is ready.

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,98 @@
+  "excludedTopics":[],
+  "numLeaderMovements":12,
+  "excludedBrokersForReplicaMove":[]
+}
```
we should be clear where this summary here, providing a full example of the `status` in the ``KafkaRebalance` resource.

tomncooper replied:
Yeah, this should be replaced with an actual copy of a rebalance status

laidan6000 replied:
Yes, good point. I'll need someone to check I've used the right format for the `status`.

To improve usability, I suggest renaming `summary` to `rebalanceproposalsummary` in the status response.

laidan6000 replied:
TO DO: Review the formatting of the rebalance status.

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+[id='proc-implementing-rebalance-proposal-{context}']
+= Implementing a rebalance proposal
+
+If you are satisfied with a rebalance proposal, you can implement it by annotating the corresponding `KafkaClusterRebalance` resource. 
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+
+If you are satisfied with a rebalance proposal, you can implement it by annotating the corresponding `KafkaClusterRebalance` resource. 
+
+At the next reconciliation, the Cluster Operator will instruct Cruise Control to rebalance the Kafka cluster. 
```
no. when the proposal is in ProposalReady state, the user has to apply `strimzi.io/rebalance=approve` annotation to trigger the actual rebalancing.

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+
+* You have xref:proc-generating-rebalance-proposals-{context}[generated a rebalance proposal].
+
+* The rebalance proposal is ready to implement (the `KafkaClusterRebalance` custom resource status is `ProposalReady`).
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+
+.Procedure
+
+. (Optional) Refresh the rebalance proposal that you want to implement:
```
we should be clear why the user should refresh so if after having the proposal ready, he doesn't execute it for long time and it becomes stale.

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+
+. (Optional) Refresh the rebalance proposal that you want to implement:
+
+.. Annotate the `KafkaClusterRebalance` resource in Kubernetes:
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl annotate kafkaclusterrebalance _cluster-name_-kafka strimzi.io/rebalance=refresh
```
`kafkarebalance`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+kubectl annotate kafkaclusterrebalance _cluster-name_-kafka strimzi.io/rebalance=refresh
+----
+
+.. Check the status of the `KafkaClusterRebalance` resource:
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl get kafka kafkaclusterrebalance -o jsonpath='{.status}'
```
`kafkarebalance`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+
+. Approve the rebalance proposal that you want to implement.
++
+Annotate the `KafkaClusterRebalance` resource in Kubernetes:
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl annotate kafkaclusterrebalance _cluster-name_-kafka strimzi.io/rebalance=approved
```
`kafkarebalance`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+kubectl annotate kafkaclusterrebalance _cluster-name_-kafka strimzi.io/rebalance=approved
+----
+
+. Wait for the next reconciliation to occur (every two minutes by default).
```
no. the rebalancing process starts immediately when the "approve" annotation is applied.

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+
+. Wait for the next reconciliation to occur (every two minutes by default).
++
+At the next reconciliation, the Cluster Operator detects the annotated resource and instructs Cruise Control to rebalance the Kafka cluster.
```
no. see above.

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
++
+At the next reconciliation, the Cluster Operator detects the annotated resource and instructs Cruise Control to rebalance the Kafka cluster.
+
+. Check the status of the `KafkaClusterRebalance` resource:
```
`KafkaRebalance`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl get kafka kafkaclusterrebalance -o jsonpath='{.status}'
```
`kafkarebalance`

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+[id='con-rebalance-goals-proposals-{context}']
+= Rebalance goals and proposals
+
+To rebalance a Kafka cluster, Cruise Control uses _rebalance goals_ to generate _rebalance proposals_. 
```
So this is a nit, but the goals are not strictly "rebalance" goals, they are "optimization" goals.

laidan6000 replied:
Changed to "optimization goals" throughout.

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+To rebalance a Kafka cluster, Cruise Control uses _rebalance goals_ to generate _rebalance proposals_. 
+
+[discrete]
+== Rebalance goals
```
Optimization Goals would be a better (more accurate) name

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+[discrete]
+== Rebalance goals
+
+Rebalance goals specify targets for resource distribution and utilization across a Kafka cluster. 
```
We are not redistributing resources (we are not moving CPUs and RAM) we are redistributing load.

laidan6000 replied:
Good point. I've updated to "workload redistribution".

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+=== Default and custom rebalance goals
+
+When interacting with Cruise Control to generate rebalance proposals, you can use the _default rebalance goals_, which are defined on deployment, or set ad hoc _custom rebalance goals_. 
+Both types of rebalance goal can be configured to include hard goals as well as soft goals.  
```
Well, you can have as many soft goals as you like but you *must* include all the configured `hard.goals` or include `skip_hard_goal_check=true`. Else the rebalance request will fail.

laidan6000 replied:
I've attempted to re-word this. This is meant to be a brief summary, with the details coming later in the table. 

Please re-read this section after I push up the changes.

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+The default rebalance goals are set in the `Kafka` custom resource. 
+You specify a list of goals in `spec.cruiseControl.config.default.goals`.
+
+Unless you set custom rebalance goals, the default rebalance goals are used by Cruise Control to return a _cached rebalance proposal_. 
```
Unless you set custom rebalance goals, in the `KafakRebalance` custom resource,...

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+You specify a list of goals in `spec.cruiseControl.config.default.goals`.
+
+Unless you set custom rebalance goals, the default rebalance goals are used by Cruise Control to return a _cached rebalance proposal_. 
+The cached proposal reflects the current state of the Kafka cluster and is updated every 15 minutes by default. 
```
The cached proposal is generated every 15 min by default using the `default.goals` specified in Kafka spec. If no default goals are specified in the Kafka spec then the `default.goals=goals`. 

Are we documenting what the default `default.goals` are?

laidan6000 replied:
Added a note:

> If no default optimization goals are specified, the cached proposal is generated using the master goals (specified in Kafka.spec.cruiseControl.config.goals).

You've raised a good point here. So far, I haven't even mentioned the `goals` list, which I think of as the "master goals".

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+
+Use custom rebalance goals to generate a rebalance proposal that addresses a particular optimization scenario. 
+For example, you might want to optimize leader replica distribution across the Kafka cluster without considering other goals, such as disk or CPU usage. 
+So, you create a `KafkaClusterRebalance` resource containing a custom rebalance goal for leader replica distribution only.
```
We do need to clearly document that if they specify a list of goals in the `KafkaRebalance` resource that does not include all the configured `hard.goals` they will get an error. In that case they need to set `skiphardgoalscheck: true` in the resource.

laidan6000 replied:
In which section of the resource? Please can you give an example configuration.

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+For example, you might want to optimize leader replica distribution across the Kafka cluster without considering other goals, such as disk or CPU usage. 
+So, you create a `KafkaClusterRebalance` resource containing a custom rebalance goal for leader replica distribution only.
+
+Using custom rebalance goals often returns a rebalance proposal faster than the default rebalance goals. 
```
I don't know where you got that idea but it is not true.

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+So, you create a `KafkaClusterRebalance` resource containing a custom rebalance goal for leader replica distribution only.
+
+Using custom rebalance goals often returns a rebalance proposal faster than the default rebalance goals. 
+They must always be a subset of the default rebalance goals.
```
This is kinda true, they must be a subset of the `goals` list.

laidan6000 replied:
True. Again, this highlights a lack of any information on the `goals` list (I'm calling these the master goals).

tomncooper commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+|===
+
+m¦JSON property
+¦Description
```
I will find some time to update the explainer docs with definitions of what the terms mean.

tomncooper commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,98 @@
+= Generating rebalance proposals 
+
+When you create or update a `KafkaClusterRebalance` resource, Cruise Control generates a _rebalance proposal_ in JSON.
+This summarizes an optimal mapping of partitions to brokers for the Kafka cluster based on the rebalance goals that were used along with statistics on the cluster rebalance.
```
So it is not just partitions to brokers, it also proposes changes to partition replica leaders.

laidan6000 replied:
Reformatted as a bulleted list.

kyguy commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+. Leader replica distribution
+. Preferred leader election
+. Intra-broker disk capacity
+. Intra-broker disk usage distribution    
```
There is still some outstanding work needed to support the CPU capacity,  CPU utilization distribution,  Intra-broker disk capacity, and Intra-broker disk usage distribution goals, so those can be removed from this list

laidan6000 replied:
OK. I've commented out those goals from the list.

laidan6000 replied:
At the start, we state that CC is used for: 

"_Monitoring the CPU_, disk, and network load of Kafka brokers."

Can that stay?

kyguy replied:
I think we can leave it as it is as a summary for CC. CC will still monitor CPU when running on k8s, it's just the values will not be accurate in all situations (at least until we get the metric reporter patch in)

kyguy commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+[discrete]
+=== Default and custom rebalance goals
+
+When interacting with Cruise Control to generate rebalance proposals, you can use the _default rebalance goals_, which are defined on deployment, or set ad hoc _custom rebalance goals_. 
```
Should we be mentioning custom goals when we don't support the addition of them?

kyguy replied:
Ah I understand, you mean the user-specified goals, not the write-your-own-goals

PaulRMellor commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -11,12 +11,19 @@
+Monitoring this broker load and reassigning busy partitions to brokers with spare capacity is time consuming and adds to the overhead of managing a Kafka cluster.
+
+Cruise Control helps to automate the monitoring and rebalancing of partitions across a Kafka cluster.
 When deployed alongside a Kafka cluster, Cruise Control can perform the following cluster balancing tasks:
```
_alongside_? In the same Kubernetes cluster?

laidan6000 replied:
Yes. Changed the wording to: "When deployed to a Kubernetes cluster that contains a Kafka cluster".

Feels a bit clumsy -- further suggestions are welcome!

PaulRMellor commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -11,12 +11,19 @@
+Monitoring this broker load and reassigning busy partitions to brokers with spare capacity is time consuming and adds to the overhead of managing a Kafka cluster.
+
+Cruise Control helps to automate the monitoring and rebalancing of partitions across a Kafka cluster.
 When deployed alongside a Kafka cluster, Cruise Control can perform the following cluster balancing tasks:
```
_can perform_? Or _performs_? Is this by default, automatic or depending on how its configured?

laidan6000 replied:
_Performs_ works better here. Thanks.

The monitoring part is automatic; the rest relies on user input.

PaulRMellor commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+You can designate certain rebalance goals as _hard goals_ that must be satisfied--for more information, see link:#soft-hard-goals[Soft goals and hard goals].
+
+With a few exceptions, cluster rebalancing in {ProductName} supports the same goals as the Kafka Cruise Control project (where the term "optimization goals" is used). 
+Following are the supported goals, which you can reorder in priority, enable, or disable:
```
```suggestion
Following are the supported goals, which you can reorder in terms of priority, and enable or disable:
```

PaulRMellor commented on documentation/modules/cruise-control/con-rebalance-goals-proposals.adoc:
```diff
@@ -0,0 +1,180 @@
+
+¦Soft goals
+
+¦Also known as "best effort" goals, soft goals do _not_ need to be satisfied in rebalance proposals. 
```
"best effort" -- I tend to only italics (_best effort_) to emphasize terms

laidan6000 replied:
Changed to italics. Don't know why I used quotation marks!

PaulRMellor commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,74 @@
+At the next reconciliation, the Cluster Operator will instruct Cruise Control to rebalance the Kafka cluster. 
+Cruise Control then reassigns partitions to brokers as summarized in the rebalance proposal.
+
+WARNING: This is not a dry run. 
```
Is _dry run_ universally understood?

laidan6000 replied:
Yes, I think it's OK from a style perspective.

laidan6000 commented:
This is now ready for re-review, thank you.

- Applied all review comments
- Added information on "master goals", which was previously missing
- Removed the tables in favour of a list of sections describing each goal configuration in turn
- Edited for clarity and consistency

laidan6000 commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,133 @@
++
+[source=yaml]
+----
+goals: {}
```
This is contradicted in the "Deploying Cruise Control" procedure, which has a list of `goals` in the CC deployment configuration.
Should I change it to an empty value in that procedure?

tomncooper replied:
Yes it should be empty. Or better yet not included at all if they are happy with the defalt master goals list.

tomncooper commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,133 @@
+[discrete]
+=== Default optimization goals
+
+We recommend that you always set _default optimization goals_ for the Kafka cluster.
```
Not sure I would be as definitive as "we recommend". I would approach this like the master goals above. By default the `default.goal` list will = the `goals` list. If however you are only concerned with a certain set of optimization goals then set them here.

tomncooper replied:
We do not recommend this. Can we take this line out but mention that they can be customised.

tomncooper commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,133 @@
+Default optimization goals are set in the Cruise Control deployment configuration: `Kafka.spec.cruiseControl.config.default.goals`
+
+The default optimization goals generate the _cached optimization proposal_ from the workload model (unless you set user-provided optimization goals in a `KafkaRebalance` custom resource). 
+This is updated to reflect the current state of the Kafka cluster every 15 minutes by default. 
```
Dunno if it is worth mentioning that this can be altered by setting the `proposal.expiration.ms` setting in the Kafka CR CC config? More often is better for fast changing clusters but will put more load on CC server.

laidan6000 replied:
Yes, I think so.
I've moved all this info to the optimization proposals concept module, and included reference to the `proposal.expiration.ms` setting.

tomncooper commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,133 @@
+* Include all configured hard goals, or an error will occur
+* Be a subset of the master optimization goals
+
+To exclude the configured hard goals from an optimization proposal, add the `skiphardgoalscheck: true` option to the `KafkaRebalance` custom resource.
```
So that option will not exclude the `hard.goal`, it will just mean that CC does not complain if you miss any of the `hard goals` out of the user supplied rebalance goals. 

This should probably read something like: "If, for some reason, you want to rebalance based on a list of goals that do not include all the configured `hard.goals` then add the `skiphardgoalscheck: true`...."

ppatierno replied:
the field is `skipHardGoalCheck`

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦numLeaderMovements
+¦Description
```
This number indicates how many partitions will have their leader switched to another replica from the current one. This is a "cheap" operation (a config change in Zookeeper).

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦excludedTopics
+¦Description
```
This is a list of topics that will be excluded for partition reassignments. Topic exclusion is not yet supported so this list will be empty.

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦excludedBrokersForReplicaMove
+¦Description
```
Some CC rebalance options allow for brokers to be excluded from the optimization proposals. These options are not currently supported so this list will be empty.

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦numIntraBrokerReplicaMovements
+¦Description
```
This indicates how many partition replicas in total will be transferred between the brokers in the Kafka Cluster.

kyguy replied:
I think this is "how many partition replicas in total will be transferred between the disks of an individual brokers, since it is "intra"  instead "inter", at least that is how the goals define the meaning, but I could be mistaken! I concede to you

If this is between disks then this isn't supported as I never added the auto configuration for JBOD!

tomncooper replied:
Oh good point! Got my definitions mixed up, thanks for checking Kyle.

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦numReplicaMovements
+¦Description
```
This number indicates how many partition replicas will be moved, including possibly between disks within the same broker.

tomncooper replied:
Correction: This number indicates how many partition replicas will be moved between seperate brokers.

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦intraBrokerDataToMoveMB
+¦Description
```
This value is derived by summing the size of each partition to be moved to a separate broker (see `numIntraBrokerReplicaMovements`).

kyguy replied:
Relates to our comments above

tomncooper replied:
Correction: This value is derived by summing the size of each partition replica to be moved including, possibly, between disks on the same broker (see `numReplicaMovements`).

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦dataToMoveMB
+¦Description
```
This value is derived by summing the size of each partition replica to be moved including, possibly, between disks on the same broker (see numReplicaMovements).

tomncooper replied:
Correction: This value is derived by summing the size of each partition replica to be moved to a separate broker (see `numIntraBrokerReplicaMovements`).

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦onDemandBalancednessScoreAfter
+¦Description
```
The `onDemandBalancednessScore` is CC's measure of the overall "balancedness" of a Kafka Cluster. It is equal to 100 minus the sum of the Balancedness Score of each violated (soft) goal. For the after score this is based on the optimization proposal.

tomncooper replied:
The code for calculating this score is here: https://github.com/linkedin/cruise-control/blob/30f89251a54317fe723edcfb06450ffae65d97e4/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlUtils.java#L517 

AFAIK it is not covered in the docs?

laidan6000 replied:
No, it's not covered in the docs.

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦onDemandBalancednessScoreBefore
+¦Description
```
The `onDemandBalancednessScore` is CC's measure of the overall "balancedness" of a Kafka Cluster. It is equal to 100 minus the sum of the Balancedness Score of each violated (soft) goal. For the before score this is based on the current cluster configuration.

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,82 @@
+¦Description
+
+m¦recentWindows
+¦Description
```
The number of metrics windows that this proposal is based on.

tomncooper commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,119 @@
+  - excludedTopics:[],
+  - numLeaderMovements:12,
+  - excludedBrokersForReplicaMove:[]
+# ...
```
@laidan6000 It might be worth you running this proceedure yourself on minikube/minishift etc. Just to get a feel for it and to add an example output which has changed slightly from above. For example the condition `type` and `status` fields have swapped. Also the docs refer to the OpProp summary as "JSON format" when the above is yaml fromat. I am not sure if K8s (or some client) is converting the JSON we give it to YAML so you should double check.

laidan6000 replied:
OK

laidan6000 replied:
I have not been able to verify that the output is correct.

tomncooper replied:
What do you mean? You couldn't get a rebalance working or that the output doesn't look like that?

laidan6000 replied:
I didn't have time to test this out.

tomncooper replied:
Ok well this output is still wrong. I will have to run a quick test to make make sure this is the correct format.

tomncooper replied:
So if you use the `kubectl describe kafkarebalance <rebalance-name>` you get a status that looks like this:
```
Status:
  Conditions:
    Last Transition Time:  2020-05-19T13:50:12.533Z
    Status:                ProposalReady
    Type:                  State
  Observed Generation:     1
  Optimization Result:
    Data To Move MB:  0
    Excluded Brokers For Leadership:
    Excluded Brokers For Replica Move:
    Excluded Topics:
    Intra Broker Data To Move MB:         0
    Monitored Partitions Percentage:      100
    Num Intra Broker Replica Movements:   0
    Num Leader Movements:                 0
    Num Replica Movements:                26
    On Demand Balancedness Score After:   81.8666802863978
    On Demand Balancedness Score Before:  78.01176356230222
    Recent Windows:                       1
  Session Id:                             05539377-ca7b-45ef-b359-e13564f1458c
```

laidan6000 replied:
Thanks - I've added the status above.

Both the steps now use the `describe` command:

`kubectl describe kafkarebalance _rebalance-name_`

tomncooper commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,119 @@
+** `ProposalReady`: The optimization proposal is ready for you to review and implement, if desired.
+The status is followed by the JSON optimization proposal.
+
+. Review the JSON optimization proposal in the `status.summary` property of the `KafkaRebalance` resource.
```
Should we add the command to do this in here?

`kubectl describe kafkarebalance.kafka.strimzi.io -n <namespace> <rebalance-cr-name>`

laidan6000 replied:
Sure. I didn't realise another command was needed.

tomncooper commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,73 @@
+
+Perform these steps for the optimization proposal that you want to implement:
+
+. If the `KafkaRebalance` status has been `ProposalReady` for some time, refresh the optimization proposal in case it has become out of date:
```
We should add a prominent **NOTE:** to this with some more explanation:

**NOTE**: If the `KafkaRebalance` status has been in the `ProposalReady` state for some time, then that proposal may no longer be applicable to the current state of the Kafka Cluster. In this case you can refresh the optimization proposal to make sure it uses the latest cluster metrics.

laidan6000 replied:
Good idea.

tomncooper commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,73 @@
+
+** NotReady: An error occurred. 
+
+//Is there a way to verify the rebalance?
```
Well you can ask for a new optimization proposal which should contain little or no changes? There is no direct way to check though.

ppatierno replied:
I think we should also explain how to "fix" a `KafkaRebalance` when an error occurs and how to move it from `NotReady` to ask for a new proposal.

laidan6000 replied:
OK. I'd prefer to document the "fix" procedure in a separate PR after this one.

tomncooper commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,73 @@
+
+. Cruise Control returns one of three statuses:
+
+** Rebalancing: The cluster rebalance operation is in progress. 
```
I may have missed it, but where is the procedure for stopping an ongoing rebalance?

laidan6000 replied:
I decided to leave that out because it seemed like an edge case. Should I add one?

tomncooper replied:
Yes definitely! Rebalances can go on for hours and seriously effect a cluster's performance. A user may mis-configure a rebalance request and not be paying attention to the summary and only realise they have made a mistake once it has begun. We need to give them an escape hatch!

ppatierno commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,133 @@
+. Leader replica distribution
+. Preferred leader election
+//. Intra-broker disk capacity
+//. Intra-broker disk usage distribution   
```
what are this line witn "//" ? this list is not clear to me. Are there titles and topics?

laidan6000 replied:
`//` is comment notation in Asciidoc. These goals are not supported as confirmed by @kyguy. I decided to comment out rather than delete.

ppatierno commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,133 @@
+
+[id="master-goals"]
+[discrete]
+=== Master optimization goals
```
from where "Master" comes from?

laidan6000 replied:
That's my choice of term. Otherwise we have _goals_, _default goals_, and _user-provided goals_, which are all different types of goals!

ppatierno replied:
Why not just using the CC community terms. If they are called goals, default goals, user-provided goals .. we have to take them not using something different tbh. @tomncooper @kyguy ?

laidan6000 replied:
"User-provided goals" is also our choice of term.

ppatierno replied:
but user-provided is just related to the fact that the user is providing them through our technology ... the `KafkaRebalance` resource, so I think that it's right.

tomncooper replied:
personally I am happy with the "master goals list" it conveys better what they are. If a goal is not in the master goal list it is not available for any CC function.

laidan6000 replied:
> but user-provided is just related to the fact that the user is providing them through our technology ... the `KafkaRebalance` resource, so I think that it's right.

Right, I see what you mean. It's more of a description than an actual term.

ppatierno commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,133 @@
+We recommend that you always set _default optimization goals_ for the Kafka cluster.
+You can include any or all of the master optimization goals in the default optimization goals.
+
+Default optimization goals are set in the Cruise Control deployment configuration: `Kafka.spec.cruiseControl.config.default.goals`
```
I think that this notation in this case is confusing. As we are using this notation it means:

```yaml
spec:
  cruiseControl:
     config:
        default:
            goals:
```

which is of course wrong because the right field is `default.goals` (so it's a field as a whole not `goals` under `default`).
For this reason I would say `default.goals` under `Kafka.spec.cruiseControl.config`.

laidan6000 replied:
Good suggestion, thanks.

ppatierno commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,133 @@
+[source]
+RackAwareGoal; ReplicaCapacityGoal; DiskCapacityGoal; NetworkInboundCapacityGoal; NetworkOutboundCapacityGoal; CpuCapacityGoal
+
+All configured hard goals _must_ be satisfied in optimization proposals (unless `skip_hard_goal_check: true` is specified in the `KafkaRebalance` custom resource).
```
`skipHardGoalCheck`

ppatierno commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,133 @@
+All configured hard goals _must_ be satisfied in optimization proposals (unless `skip_hard_goal_check: true` is specified in the `KafkaRebalance` custom resource).
+In other words, an optimization proposal that does _not_ satisfy all the hard goals is rejected by Cruise Control and not sent to the user.
+
+Hard goals are controlled in the Cruise Control deployment configuration: `Kafka.spec.cruiseControl.config.hard.goals`
```
the same as before, `hard.goals` is a property as a whole, we should clarify as I already mentioned for `default.goals`

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,119 @@
++
+[source,shell,subs="+quotes"]
+----
+oc apply -f _your-file_
```
`kubectl`

ppatierno commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,119 @@
+  conditions:
+  - lastTransitionTime: 2020-05-23T23:46:57+0000
+    status: "True"
+    type: Ready
```
it's changed, now `type: State` (is always in this way) and `status: Ready`

scholzj commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,73 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl annotate kafkarebalance _cluster-name_-kafka strimzi.io/rebalance=refresh
```
The name of the `KafkaRebalance` resources AFAIU can be anything. So the example commands like this should not use the `_cluster-name_-kafka` as a name. They should use something generic like `_kafka-rebalance-name_` instead. Otherwise this would be misleading given the other namesfor example in example do not follow this pattern anyway. (this is not just here but on more places)

laidan6000 replied:
Good point. I'll change the variable to `_cluster-rebalance-name_` throughout.

tombentley commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,73 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl annotate kafkarebalance _cluster-name_-kafka strimzi.io/rebalance=approved
```
it's `approve` not `approved`

laidan6000 replied:
Changed it, thanks.

laidan6000 commented:
This is now ready for final checks.

* Applied reviewers' comments.
* Added new procedure for stopping a cluster rebalance

Please can we defer the procedure to "fix" a `KafkaRebalance` when an error occurs until a later PR?

There is one open query: https://github.com/strimzi/strimzi-kafka-operator/pull/2956#discussion_r425275967

ppatierno commented:
@laidan6000 I don't think we can have another PR for "fixing" the KR. Another PR means more time loss for merging it, waiting for Travis on PR, waiting for Travis for master, etc etc . I think we can add this procedure here. If you want I can write it.

tomncooper commented on documentation/modules/con-configuring-user-operator.adoc:
```diff
@@ -9,7 +9,7 @@ User Operator deployment can be configured using additional options inside the `
 
 `watchedNamespace`::
-The Kubernetes namespace in which the topic operator watches for `KafkaUsers`.
+The Kubernetes namespace in which the user operator watches for `KafkaUsers`.
```
Why is this here?

laidan6000 replied:
It's an unrelated typo.

ppatierno replied:
@laidan6000 which should not be in unrelated PRs :-) ... anyway it's ok for this time! :P

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,94 @@
+An _optimization proposal_ is a summary of proposed changes that would produce a more balanced Kafka cluster, with partition workloads distributed more evenly among the brokers. 
+Each optimization proposal is based on the set of xref:con-optimization-goals-{context}[optimization goals] that were used to generate it.
+
+Optimization proposals always satisfy all the configured hard goals, although not necessarily all the soft goals. Hard goals are ignored if `skipHardGoalCheck: true` is specified in the `KafkaRebalance` custom resource.
```
This statement is not correct. The `skipHardGoalCheck: true` only applies to the check that is done for the user defined goals and weather they contain all the specified `hard.goals`. If a user sets `skipHardGoalCheck: true` all the user defined goals that are in the `hard.goals` list will be treated as hard goals.

tomncooper replied:
You can just remove the second sentence.

laidan6000 replied:
Right. I'm going to document this in the "Hard goals" section instead.

laidan6000 replied:
I added a note.

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,94 @@
+
+Optimization proposals always satisfy all the configured hard goals, although not necessarily all the soft goals. Hard goals are ignored if `skipHardGoalCheck: true` is specified in the `KafkaRebalance` custom resource.
+
+An optimization proposal contains the following information, in JSON, about the pending cluster rebalance:
```
So an optimization proposal does contain the information in the bullet points below but we don't currently show this information to a user. This requires the `--verbose` flag which we currently don't support. We may add this in future but there is currently no way to show this information to the user.

tomncooper replied:
You could say "An optimisation proposal consists of the following information........however a summary of the changes is displayed in the KafkaRebalance CR status."

laidan6000 replied:
Removed the bullet points.

laidan6000 replied:
Please can you check this again, @tomncooper ?

tomncooper replied:
This section has changes so there is nothing to check. You can resolve this.

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,94 @@
+¦Description
+
+m¦numIntraBrokerReplicaMovements
+¦The total number of partition replicas that will be transferred between the disks of an individual broker.
```
"The total number of partition replicas that will be transferred between the disks of the cluster's brokers."

tomncooper commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,124 @@
+    strimzi.io/cluster: my-cluster
+spec:    
+  goals:
+    - RackAwareGoal,
```
Do we need a "," at the end of these names? @kyguy does this YAML list get parsed into a Java list and the "," added?

ppatierno replied:
No, this is the `KafkaRebalance` resource. `goals` is an array, so no comma.

laidan6000 replied:
The commas are not there in `examples/cruise-control`.

I can remove them from here.

tomncooper commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,124 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl get kafka kafkarebalance -o jsonpath='{.status}'
```
This command is wrong. It should be `kubectl get kafkarebalances <rebalance-name> -o jsonpath='{.status}'`

tomncooper replied:
Or better yet: `kubectl describe kafkarebalance <rebalance-name>`

tomncooper replied:
You can get the exact status output with `kubectl get kafkarebalances <rebalance-name> -o jsonpath='{.status.conditions[0].status}'` but there is no absolute guarantee that the correct condition will always be at position zero.

tomncooper commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,124 @@
+** `ProposalReady`: The optimization proposal is ready for review and, if desired, implementation.
+The status contains the optimization proposal as JSON, in the `status.summary` property of the `KafkaRebalance` resource.
+
+. Review the JSON optimization proposal.
```
It is not JSON formatted if you use the `kubectl describe` command

laidan6000 replied:
OK. I've just removed the two mentions that optimization proposals are formatted as JSON.

tomncooper commented on documentation/modules/cruise-control/proc-generating-rebalance-proposals.adoc:
```diff
@@ -0,0 +1,124 @@
+    type: State
+  observedGeneration: 4
+  summary:
+  - numIntraBrokerReplicaMovements:0,
+  - excludedBrokersForLeadership:[],
+  - numReplicaMovements:31,
+  - onDemandBalancednessScoreAfter:84.08179174145127,
+  - onDemandBalancednessScoreBefore:79.74705957325753,
+  - intraBrokerDataToMoveMB:0,
+  - recentWindows:1,
+  - dataToMoveMB:0,
+  - monitoredPartitionsPercentage:100.0,
+  - excludedTopics:[],
+  - numLeaderMovements:12,
+  - excludedBrokersForReplicaMove:[]
```
This is covered in another section, @laidan6000 can we add a ref to that section here?

tomncooper commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,161 @@
+[source]
+RackAwareGoal; ReplicaCapacityGoal; DiskCapacityGoal; NetworkInboundCapacityGoal; NetworkOutboundCapacityGoal; CpuCapacityGoal
+
+Unless `skipHardGoalCheck: true` is specified in the `KafkaRebalance` custom resource, all configured hard goals _must_ be satisfied in optimization proposals.
+In other words, an optimization proposal that does _not_ satisfy all the hard goals is rejected by Cruise Control and not sent to the user.
```
Ok so this is a confusion that has popped up a few times so we really need to make it clear. `skipHardGoalsCheck: true` does not disable hard goals or treat all goals as soft, it simply stops CC rejecting a rebalance request with a user defined goals list that doesn't contain all the defined `hard.goals` (in the CC server config). If some of the user defined goals in a rebalance with `skipHardGoalsCheck: true` happen to be in the `hard.goals` list, then they will be treated as hard goals.

tombentley replied:
OK, thank you for correcting me @tomncooper!

ppatierno replied:
@laidan6000 so we should change here as Tom suggested.

laidan6000 replied:
I've added a note to this module explaining "skipHardGoalCheck: true" in more detail. I'll leave the text as it is.

ppatierno commented on documentation/assemblies/cruise-control/assembly-cruise-control-concepts.adoc:
```diff
@@ -5,14 +5,22 @@
-This chapter provides an overview of the Kafka cluster rebalancing features provided by the {CruiseControlProject} open source project and helps you get started using Cruise Control with {ProductName}.
+You can deploy {CruiseControlProject} and access a subset of its features through {ProductName} custom resources.
+
+Example YAML files for deploying Cruise Control and setting optimization goals are provided in `examples/kafka/`.
```
Good catch!

laidan6000 commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -7,16 +7,19 @@
+Cruise Control automates the cluster rebalancing process.
+It constructs a _workload model_ of resource utilization for the cluster--based on CPU, disk, and network load--and generates _optimization proposals_ for balanced partition assignments. A set of configurable _optimization goals_ is used to calculate these proposals.
+
+You can choose to implement optimization proposals to achieve a more balanced Kafka cluster.  
```
"Apply" seems rather too generic to me. The other strong candidate was "execute".

Agree that the sentence above is a bit woolly.

laidan6000 replied:
Changed to:

> You can choose to apply those optimization proposals to your Kafka cluster so the broker pods are used most effectively, leading to a more balanced cluster.

laidan6000 commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,161 @@
+
+To ignore the configured hard goals in an optimization proposal, add the `skipHardGoalCheck: true` option to the `KafkaRebalance` custom resource.
+
+[id="hard-soft-goals"]
+[discrete]
+=== Hard goals and soft goals
+
+In Cruise Control, six of the master optimization goals are preset as _hard goals_:
+
+[source]
+RackAwareGoal; ReplicaCapacityGoal; DiskCapacityGoal; NetworkInboundCapacityGoal; NetworkOutboundCapacityGoal; CpuCapacityGoal
+
+Unless `skipHardGoalCheck: true` is specified in the `KafkaRebalance` custom resource, all configured hard goals _must_ be satisfied in optimization proposals.
+In other words, an optimization proposal that does _not_ satisfy all the hard goals is rejected by Cruise Control and not sent to the user.
+
+Hard goals are controlled in the Cruise Control deployment configuration, by editing the `hard.goals` property in `Kafka.spec.cruiseControl.config`.
+
+* To inherit the six preset hard goals from Cruise Control, do not specify the `hard.goals` property in `Kafka.spec.cruiseControl.config`
+
+* To change the preset hard goals, specify the desired goals in the `hard.goals` configuration option.
+
+Increasing the number of hard goals will reduce the likelihood of Cruise Control generating valid optimization proposals.
+
+Goals not designated as hard goals are treated as _soft goals_.
+Also known as _best effort_ goals, soft goals do _not_ need to be satisfied in optimization proposals, but are included in optimization calculations.
+An optimization proposal that violates one or more soft goals, but satisfies all hard goals, is valid.
+
+NOTE: For example, you might have a soft goal to distribute a topic's replicas evenly across the cluster (the topic replica distribution goal). 
+Cruise Control will ignore this goal if doing so enables all the configured hard goals to be met.
```
Originally, it was explained first. Honestly, I've gone through so many different iterations now, I'm not sure that a logical order even exists!

I'll shift the hard and soft goals to the top of the module and we'll see how it looks.

laidan6000 commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,161 @@
+[discrete]
+=== Hard goals and soft goals
+
+In Cruise Control, six of the master optimization goals are preset as _hard goals_:
```
Yes, that's improved.

laidan6000 commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,78 @@
+
+[id='proc-implementing-optimization-proposal-{context}']
+
+= Implementing an optimization proposal
+
+You can implement an xref:con-optimization-proposals-{context}[optimization proposal] if its status is `ProposalReady`.
```
Well, I think we'll have to stick to "implement" unless there is strong agreement on the new term.

ppatierno replied:
@laidan6000 I am not a native English but tbh "implement" doesn't sound the right term here. I agree with @tombentley about "apply". Of course, I could be wrong due to the fact I am not English. @tomncooper wdyt?

tomncooper replied:
Yeah "Implement" feels wrong as it implies that the user is going to do all the steps, when actually they are just telling CC to go and do the steps. TBF "apply" suffers from the same issue. I think the problem is wrapping this in one term. It is a two step process. The user is not "Implementing an optimization proposal" they are "Obtaining and approving an optimization proposal". 

So better to say "You can _approve_ an optimization proposal if its status is `ProposalReady`."

laidan6000 replied:
@ppatierno - You are definitely _not_ wrong. Please, never think that your suggestions on wording and terminology are less valid because you're a non-native English speaker.

I used "implement" to mean "carrying out" or "putting into effect" the actions summarized in the optimization proposal. It's a synonym of "execute". 

https://www.collinsdictionary.com/dictionary/english/implement

"Apply" is a more general verb, but it can also mean "to carry out". I just wanted to use a more specific verb in this case, to add variety to the documentation as much as anything else.

tomncooper replied:
I still think the issue is that the user is not "implementing/applying/executing" the optimisation proposal, they are allowing CC to do it. If we are being really picky a user could "implement" the OP by accessing the verbose OP (not yet supported) and getting the partition reassignment command json and applying it themselves.

ppatierno replied:
From one of the previous chapter ... "You can choose to apply those optimization proposals" ... so I see you used "apply"

laidan6000 replied:
Well, the change above was from Tom's review. But, OK, I see your point. I'll change to "apply".

Good discussion all!

laidan6000 replied:
Sorry, I re-read Tom C's comments. "Obtaining and approving an optimization proposal" might be the best choice here.

laidan6000 replied:
I changed the title to "Approving an optimization proposal". All references to "implementing" proposals elsewhere in the doc have been changed to "approving".

laidan6000 commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,78 @@
+*This is not a dry run.* Before you implement an optimization proposal, you must:
+
+* Refresh the proposal in case it has become out of date.
+* Carefully review the contents of the proposal.
```
I've added a link to "Contents of optimization proposals". Good suggestion.

I think the extra conceptual information on when to reject a proposal will have to wait. We can capture it as an issue?

laidan6000 replied:
Are there one or two main reasons why a user would reject an optimization proposal?

tomncooper replied:
There are many possible reasons. The main ones might include a really huge amount of data moving or lots of **Inter**Broker replica movements, but I think a true explanation would be quite in depth. It might be better to get the upstream CC community to improve their wiki with this kind of information. I have a couple of WIP PR on the CC docs so I can add this discussion to them.

laidan6000 replied:
OK, thanks. I think we should capture this discussion in a new Docs issue. Please add me to your CC PRs!

tomncooper commented on documentation/modules/cruise-control/con-cruise-control-overview.adoc:
```diff
@@ -7,16 +7,19 @@
+Cruise Control automates the cluster rebalancing process.
+It constructs a _workload model_ of resource utilization for the cluster--based on CPU, disk, and network load--and generates _optimization proposals_ for more balanced partition assignments. A set of configurable _optimization goals_ is used to calculate these proposals.
+
+You can choose to apply those optimization proposals to your Kafka cluster so the broker pods are used most effectively, leading to a more balanced cluster.
```
do you mean "more" effectively?

laidan6000 replied:
Adverbs are useful sometimes...

tomncooper commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,85 @@
+An _optimization proposal_ is a summary of proposed changes that would produce a more balanced Kafka cluster, with partition workloads distributed more evenly among the brokers. 
+Each optimization proposal is based on the set of xref:con-optimization-goals-{context}[optimization goals] that were used to generate it.
+
+Use the summary to decide whether to implement the cluster rebalance, or change the optimization goals. 
```
"Use the summary (shown in the Status of the KafkaRebalance resource)....."

tomncooper commented on documentation/modules/cruise-control/proc-fixing-problems-with-kafkarebalance.adoc:
```diff
@@ -0,0 +1,49 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl describe kafkarebalance.kafka.strimzi.io -n <namespace> <rebalance-cr-name>
```
We should be consistent and use just `kafkarebalance` here

tomncooper replied:
Also why are you using namespace here when none of the other command examples specify it? Probably best to remove it and be consistent.

tomncooper commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,78 @@
+
+Perform these steps for the optimization proposal that you want to implement:
+
+. If the `KafkaRebalance` status has been `ProposalReady` for some time, it might no longer apply to the current state of the Kafka cluster. In this case, you must refresh the optimization proposal to make sure it uses the latest cluster metrics:
```
I think this is a good point. Ideally they would approve the proposal within a few mins of requesting it.

laidan6000 replied:
Agree that "some time" is ambiguous. I'll use your text.

tomncooper commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,78 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl get kafka kafkarebalance -o jsonpath='{.status}'
```
You should make all the "Check the status of the resource" command consistent. You uses `kubectl describe` in others?

laidan6000 replied:
Yes, good point. I'll check them all.

tomncooper commented on documentation/modules/cruise-control/proc-stopping-cluster-rebalance.adoc:
```diff
@@ -0,0 +1,43 @@
+
+If you want to stop a cluster rebalance operation that is in progress, apply a `stop` annotation to the `KafkaRebalance` custom resource. 
+This causes Cruise Control to stop rebalancing after the current batch of partition reassignments has completed. 
+Any partition reassignments before this point will have been successfully applied, so the Kafka cluster will be in a different state compared to before the rebalance operation started.
```
That is a good **NOTE** to add here.

ppatierno commented on documentation/modules/cruise-control/proc-stopping-cluster-rebalance.adoc:
```diff
@@ -0,0 +1,43 @@
+
+= Stopping a cluster rebalance
+
+Once started, a cluster rebalance operation might take some time to complete and affect the overall performance of the Kafka cluster.  
```
I think that adding this information opens a quite new chapter and much more work to do. We can do that but why we should explain this in Strimzi doc when it's more Kafka itself related?

laidan6000 replied:
> Above all we need to explain the difference between changing leadership of replicas (cheap) and, moving replicas between log dirs on a single broker (expensiveish) reassigning replicas between brokers (expensive).

I can add a brief note about the relative "expense" of each operation to the "contents of optimization proposals" table. Of course, there is more to your suggestion than that.

laidan6000 replied:
@tombentley - I've pushed up those changes. Please can you check the table in "Optimization proposals overview"?

tombentley commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,91 @@
+¦Not yet supported. An empty list is returned.
+
+m¦numReplicaMovements
+¦The number of partition replicas that will be moved between separate brokers.
```
@tomncooper are we confident that this is interbroker movement, and not the sum of interbroker and intrabroker?

tomncooper replied:
I will double check the CC source code, just to make sure.

tomncooper replied:
That number comes from [this method](https://github.com/linkedin/cruise-control/blob/e5fdfe8df81957e69fd50b9e290e72e45d26c812/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizerResult.java#L232) and is a count incremented in an `if` branch for removal or addition of replicas to a broker which is separate from the `if` branch which increments the **intra**broker (between disks) count, so I pretty confident this is the **inter**broker replica movement count.

laidan6000 replied:
OK, I'll leave this unchanged.

tombentley commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,91 @@
+¦The number of metrics windows upon which the optimization proposal is based.
+
+m¦dataToMoveMB
+¦The sum of the size of each partition replica that will be moved to a separate broker (see also `numReplicaMovements`).
```
I guess it might be worth calling out that is (and to a lesser extent `intraBrokerDataToMoveMB`) also provide a guide about how costly the balancing is going to be.

tomncooper replied:
That is a good idea.

laidan6000 replied:
Sorry, I don't quite understand your suggestion - @tombentley

tomncooper replied:
The bigger this number is (and to a lesser extent `intraBrokerDataToMoveMB`), the longer a rebalance will take .

ppatierno commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -0,0 +1,164 @@
+    config:
+      default.goals: >
+         com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,
+         com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal
```
comma missing at the end

ppatierno commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -0,0 +1,91 @@
+An _optimization proposal_ is a summary of proposed changes that would produce a more balanced Kafka cluster, with partition workloads distributed more evenly among the brokers. 
+Each optimization proposal is based on the set of xref:con-optimization-goals-{context}[optimization goals] that were used to generate it.
+
+Use the summary (shown in the `Status` of the `KafkaRebalance` resource) to decide whether to implement the cluster rebalance, or change the optimization goals.
```
`status`

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,79 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl annotate kafkarebalance _cluster-rebalance-name_ strimzi.io/rebalance=refresh
```
`_cluster-rebalance-name_` should be `_rebalance-cr-name_` as other commands.

ppatierno commented on documentation/modules/cruise-control/proc-implementing-rebalance-proposal.adoc:
```diff
@@ -0,0 +1,79 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl annotate kafkarebalance _cluster-rebalance-name_ strimzi.io/rebalance=approve
```
ditto as above, use `_rebalance-cr-name_`

ppatierno commented on documentation/modules/cruise-control/proc-stopping-cluster-rebalance.adoc:
```diff
@@ -0,0 +1,45 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl annotate kafkarebalance _cluster-rebalance-name_ strimzi.io/rebalance=stop
```
ditto as above, use `_rebalance-cr-name_`

laidan6000 commented:
Thank you for the thorough reviews. I've applied all the comments apart from those that are out of scope or need further discussion.

There are 3 discussions still open. @tomncooper and @tombentley - Please can you check them and approve the PR if you're happy?

@ppatierno - I actioned all your most recent comments.
